### PR TITLE
[FIX] pos_gift_card : Can sell the same gift card multiple times

### PR DESCRIPTION
**Pull request by robinengels**
Current behavior:
When creating a gift card in the backend you were able to sell it multiple times in a PoS

Steps to reproduce:
- Create a gift card in the backend
- Open PoS with gift card mode in "Scan a barcode and set the price" or "Scan an existing barcode with an existing price"
- You can sell this same card multiple times

opw-2666917
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/pos_gift_card/static/src/js/GiftCardPopup.js [hunk 1]
```diff
@@ -4,6 +4,9 @@ odoo.define("pos_gift_card.GiftCardPopup", function (require) {
   const { useState, useRef } = owl.hooks;
   const AbstractAwaitablePopup = require("point_of_sale.AbstractAwaitablePopup");
   const Registries = require("point_of_sale.Registries");
+  var core = require('web.core');
```
Comment (kitan191) [inline addons/pos_gift_card/static/src/js/GiftCardPopup.js:7]: You can use `const`

File: addons/pos_gift_card/static/src/js/GiftCardPopup.js [hunk 2]
```diff
@@ -42,12 +45,21 @@ odoo.define("pos_gift_card.GiftCardPopup", function (require) {
...
+          price: this.state.amountToSet,
+          quantity: 1,
+          merge: false,
+          generated_gift_card_ids: giftCard ? giftCard.id : false,
+        });
+      }
+      else {
+        Gui.showPopup('ErrorPopup', {
+          'title': _t('This gift card has already been sold'),
+          'body': _t('You cannot sell a gift card that has already been sold')
```
Comment (kitan191) [inline addons/pos_gift_card/static/src/js/GiftCardPopup.js:60]: If you add translatable terms, you should add the corresponding entries in the pot file.

File: addons/pos_gift_card/static/src/js/GiftCardPopup.js [hunk 3]
```diff
@@ -62,10 +74,10 @@ odoo.define("pos_gift_card.GiftCardPopup", function (require) {
             model: "gift.card",
             method: "search_read",
             args: [[["code", "=", this.state.giftCardBarcode]]],
-            fields: ["code", "initial_amount", "balance"],
+            fields: ["code", "initial_amount", "balance","buy_pos_order_line_id"],
```
Comment (kitan191) [inline addons/pos_gift_card/static/src/js/GiftCardPopup.js:77]: Missing space after the comma.

File: addons/pos_gift_card/static/src/js/models.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ odoo.define("pos_gift_card.gift_card", function (require) {
   models.load_models([
     {
       model: "gift.card",
-      fields: ["code", "initial_amount", "balance"],
+      fields: ["code", "initial_amount", "balance","buy_pos_order_line_id"],
```
Comment (kitan191) [inline addons/pos_gift_card/static/src/js/models.js:19]: Missing space after the comma.

File: addons/pos_gift_card/static/src/js/GiftCardPopup.js [hunk 5]
```diff
@@ -62,10 +74,10 @@ odoo.define("pos_gift_card.GiftCardPopup", function (require) {
...
           });
           if (giftCard.length) {
-            this.env.pos.giftCard.push(giftCard[0])
+            this.env.pos.giftCard.push(giftCard[0]);
```
Comment (kitan191) [inline addons/pos_gift_card/static/src/js/GiftCardPopup.js:80]: Don't add a missing semi-colon on a line that is not part of your fix.
It will add unnecessary diff.

**Comment by robinengels** [comment]
@pimodoo

File: addons/pos_gift_card/static/src/js/GiftCardPopup.js [hunk 6]
```diff
@@ -37,17 +39,34 @@ odoo.define("pos_gift_card.GiftCardPopup", function (require) {
...
+        });
+      } else {
+        await this.showPopup('ErrorPopup', {
+          'title': _t('This gift card has already been sold'),
+          'body': _t('You cannot sell a gift card that has already been sold'),
+        });
+      }
     }
 
     async getGiftCard() {
```
Comment (pimodoo) [inline addons/pos_gift_card/static/src/js/GiftCardPopup.js:72]: it seems to make a lot of change, and maybe we can ease it. Can't we just change this "GetGiftCard" function, to get all fields. And so after that we know that if there are buy lines and/or pos buy lines? We'll be able to know if it has already been sold in the js.

Comment (robinengels) [reply addons/pos_gift_card/static/src/js/GiftCardPopup.js:72]: I tried to do it this way the first time but it was not working. The getGiftCard method is only retrieving the gift card from the server if it's not already in the PoS. Wich is the case if the gift card was created before the opening of the PoS so the version of the gift card is not the right one. However, we could change the getGiftCard to update the gift card if it's already in the PoS. What do you think is the best solution?

Comment (pimodoo) [reply addons/pos_gift_card/static/src/js/GiftCardPopup.js:72]: It is because of:
"      let giftCard = this.env.pos.giftCard.find(
        (gift) => gift.code === this.state.giftCardBarcode
      );
 "
in getGiftCard, which allow the use of gift cart "Offline", but as you want to check the validity on the server each tim, you can just remove it, remove the load of existing giftcards in models, and jus let the search_read, without fields to get all of them

File: addons/pos_gift_card/static/src/js/GiftCardPopup.js [hunk 7]
```diff
@@ -42,35 +44,38 @@ odoo.define("pos_gift_card.GiftCardPopup", function (require) {
...
-      this.env.pos.get_order().add_product(gift, {
-        price: this.state.amountToSet,
-        quantity: 1,
-        merge: false,
-        generated_gift_card_ids: giftCard ? giftCard.id : false,
-      });
+
+      let can_be_sold = !giftCard.buy_pos_order_line_id && !giftCard.buy_line_id;
+
+      if (can_be_sold || !giftCard) {
```
Comment (pimodoo) [inline addons/pos_gift_card/static/src/js/GiftCardPopup.js:50]: Why is ther !giftCard ? it means that when you generate a new one, it won't work. more over in the generated_gift_card_ids, you have "giftCard ? giftCard.id : false,"

Comment (robinengels) [reply addons/pos_gift_card/static/src/js/GiftCardPopup.js]: It was working with it but it was actually not needed. When the gift card is False `can_be_sold` will always be true so the `!giftCard` was not needed.

File: addons/pos_gift_card/static/src/js/GiftCardPopup.js [hunk 8]
```diff
@@ -37,40 +39,43 @@ odoo.define("pos_gift_card.GiftCardPopup", function (require) {
...
           this.env.pos.config.gift_card_product_id[0]
         ];
-      this.env.pos.get_order().add_product(gift, {
-        price: this.state.amountToSet,
-        quantity: 1,
-        merge: false,
-        generated_gift_card_ids: giftCard ? giftCard.id : false,
-      });
+
+      let can_be_sold = !giftCard.buy_pos_order_line_id && !giftCard.buy_line_id;
```
Comment (pimodoo) [inline addons/pos_gift_card/static/src/js/GiftCardPopup.js:48]: don't you get an error when there is no giftcard passed? Because False doesn't have buy_pos_order_line_id attribute

Comment (robinengels) [reply addons/pos_gift_card/static/src/js/GiftCardPopup.js:48]: No everything is working properly (https://watch.screencastify.com/v/nU0YXZJleD80W0bPjsv6)

**Comment by pimodoo** [comment]
@robinengels it looks good to me, but as we are not using the local gift cards anymore, could you remove the loading of them in models.js and the different occurence of calls to pos.giftCard

File: addons/pos_gift_card/static/src/js/GiftCardPopup.js [hunk 9]
```diff
@@ -4,6 +4,8 @@ odoo.define("pos_gift_card.GiftCardPopup", function (require) {
   const { useState, useRef } = owl.hooks;
   const AbstractAwaitablePopup = require("point_of_sale.AbstractAwaitablePopup");
   const Registries = require("point_of_sale.Registries");
+  const core = require('web.core');
+  const _t = core._t;
```
Comment (caburj) [inline addons/pos_gift_card/static/src/js/GiftCardPopup.js:8]: No need to import this. You have access to `_t` in every owl.Component via `this.env`.

**Review by caburj — APPROVED** [review]

**Comment by pimodoo** [comment]
@robodoo r+ rebase-ff
